blocks/hooks/custom-styles.js

```diff
@@ -0,0 +1,69 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { Component, getWrapperDisplayName } from '@wordpress/element';
+import { addFilter } from '@wordpress/hooks';
+
+/**
+ * Internal dependencies
+ */
+import { getBlockRandomAnchor, hasBlockSupport, getBlockType } from '../api';
+
+const CUSTOM_STYLE_ATTRIBUTES = [ 'backgroundColor', 'textColor' ];
+
+const hasCustomStyles = ( attributes ) => (
+	CUSTOM_STYLE_ATTRIBUTES.some( attr => attributes[ attr ] )
+);
+
+class AddAnchorWhenNeeded extends Component {
+	componentWillReceiveProps( nextProps ) {
+		if ( ! nextProps.attributes.anchor && hasCustomStyles( nextProps.attributes ) ) {
+			nextProps.setAttributes( {
+				anchor: getBlockRandomAnchor( nextProps.name ),
+			} );
+		}
+	}
+
+	render() {
+		return null;
+	}
+}
+
+export function addAnchorWhenStylesNeed( BlockEdit ) {
+	const WrappedBlockEdit = ( props ) => {
```

Comment:
@jorgefilipecosta you might want to combine `AddAnchorWhenNeeded` and `WrappedBlockEdit` together and provide only one `Component`. It looks like `AddAnchorWhenNeeded` is there only for lifecycle method, so could be also a part of `WrappedBlockEdit`. This would simplify render method and help keep `componentWillReceiveProps` focused on logic related to anchor.

Reply:
Thank you for rebasing and updating this PR @gziolo. I tried to simplify things a bit following your advice :)